Feature/install from folder

bosonie opened the pull request:
Fixes #79
Fixes #81

bosonie commented:
@sphuber I implemented a fix that touches the current code as little as possible. However it is worth it a more consistent refactoring, meaning modifications to the `create_family_from_archive`. What do you think?
Regarding the CI fails. It is for the pylint `consider-using-with`, however I do not think that nesting `with` commands is the way. So I leave it like that and fix it afer your opinion on my implementation.
I also fixed the CI problem every packaged had (pinned postgresql version), but if you prefer, I can remove it and do a separate PR.

bosonie commented:
I'm also noticing now that 
```
aiida-pseudo install family https://legacy-archive.materialscloud.org/file/2018.0001/v4/SSSP_1.0_PBE_efficiency.tar.gz label
```
that should be tested in `tests/cli/test_install.py` does actually fail in real life. Not of because of a problem in the CLI command, but for failing of the underline `family_type.create_from_folder`. Is it expected?

bosonie commented:
@sphuber @mbercx This is ready for review.

sphuber commented:
> I'm also noticing now that
> 
> ```
> aiida-pseudo install family https://legacy-archive.materialscloud.org/file/2018.0001/v4/SSSP_1.0_PBE_efficiency.tar.gz label
> ```
> 
> that should be tested in `tests/cli/test_install.py` does actually fail in real life. Not because of a problem in the CLI command, but for failing of the underline `family_type.create_from_folder`. Is it expected?

How is it failing exactly? And is it failing on the `master` branch or with your patch? Or both?

sphuber commented on aiida_pseudo/cli/params/options.py:
```diff
@@ -91,3 +91,12 @@
     )
 )
+
+PSEUDOS_TYPE = OverridableOption(
```
```suggestion
PSEUDO_TYPE = OverridableOption(
```

sphuber commented on aiida_pseudo/cli/params/options.py:
```diff
@@ -91,3 +91,12 @@
+
+PSEUDOS_TYPE = OverridableOption(
+    '-P',
+    '--pseudos-type',
```
```suggestion
    '--pseudo-type',
```

sphuber commented on aiida_pseudo/cli/params/options.py:
```diff
@@ -91,3 +91,12 @@
+    type=PseudoPotentialTypeParam(),
+    default='pseudo',
+    show_default=True,
+    help='Choose the type of pseudo potentials belonging to the family.'
```
```suggestion
    help='Select the pseudopotential type to be used for the family. Should be the entry point name of a subclass of `PseudoPotentialData`.'
```

sphuber commented on aiida_pseudo/cli/params/types.py:
```diff
@@ -5,7 +5,45 @@
+        if not issubclass(pseudo_type, PseudoPotentialData):
+            raise click.BadParameter(f'`{value}` entry point is not a subclass of `PseudoPotentialData`.')
+
+        PseudoPotentialData.entry_point = value
```
I noticed that this line is also there for `PseudoPotentialFamilyParam` but I don't think it does anything and so is better to be removed

bosonie replied:
This would make fail the tests `tests/cli/test_list.py::test_list_filter`and `tests/cli/test_list.py::test_list_filter_no_result`

sphuber replied:
Ah I remember now, I added the original value that the user specifies on the param, because the converted value will be the class that correspond to the plugin. However, in the actual `list` command, I need the original entry point. Ok, good to keep then! :smile:

sphuber commented on aiida_pseudo/cli/params/types.py:
```diff
@@ -5,7 +5,45 @@
+class PseudoPotentialTypeParam(click.ParamType):
+    """Parameter type for `click` commands to define a subclass of `PseudoPotentialData`."""
+
+    name = 'pseudos_type_type'
```
```suggestion
    name = 'pseudo_type'
```

sphuber commented on aiida_pseudo/cli/params/types.py:
```diff
@@ -5,7 +5,45 @@
+    def convert(self, value, _, __):
+        """Convert the entry point name to the corresponding class.
+
+        :param value: entry point name that should correspond to subclass of `PseudoPotentialFamily` group plugin
```
```suggestion
        :param value: entry point name that should correspond to subclass of `PseudoPotentialData` data plugin
```

sphuber commented on aiida_pseudo/cli/params/types.py:
```diff
@@ -5,7 +5,45 @@
+        """Convert the entry point name to the corresponding class.
+
+        :param value: entry point name that should correspond to subclass of `PseudoPotentialFamily` group plugin
+        :return: the `PseudoPotentialFamily` subclass
```
```suggestion
        :return: the `PseudoPotentialData` subclass
```

sphuber commented on aiida_pseudo/cli/params/types.py:
```diff
@@ -5,7 +5,45 @@
+
+        :param value: entry point name that should correspond to subclass of `PseudoPotentialFamily` group plugin
+        :return: the `PseudoPotentialFamily` subclass
+        :raises: `click.BadParameter` if the entry point cannot be loaded or is not subclass of `PseudoPotentialFamily`
```
```suggestion
        :raises: `click.BadParameter` if the entry point cannot be loaded or is not subclass of `PseudoPotentialData`
```

sphuber commented on tests/cli/test_install.py:
```diff
@@ -129,6 +130,26 @@ def test_install_family(run_cli_command, get_pseudo_archive):
 
 
+pytest.mark.usefixtures('clear_db')
+
+
```
```suggestion
```

sphuber replied:
I think there are still unnecessary empty lines here

sphuber commented on tests/cli/test_install.py:
```diff
@@ -13,6 +13,7 @@
 from aiida_pseudo.groups.family import PseudoPotentialFamily
 from aiida_pseudo.groups.family.pseudo_dojo import PseudoDojoFamily, PseudoDojoConfiguration
 from aiida_pseudo.groups.family.sssp import SsspFamily, SsspConfiguration
+from aiida_pseudo.data.pseudo.upf import UpfData
```
I'm trying to keep alphabetical order for imports

sphuber commented on tests/cli/test_install.py:
```diff
@@ -129,6 +130,26 @@ def test_install_family(run_cli_command, get_pseudo_archive):
+
+    result = run_cli_command(cmd_install_family, options)
+    assert f'installed `{label}`' in result.output
+    assert PseudoPotentialFamily.objects.count() == 2
```
why should this be 2? The database is cleared before this test isn't it?

sphuber replied:
Still don't really understand why this should be 2? Maybe because this test is not properly decorated (because of the empty lines after the decorator with the fixture `clean_db`)

bosonie replied:
Sorry I forgot to check about this. Will do now. All the rest has been addressed.

bosonie replied:
The decorator was misplaced. I'll push now the fix. Sorry!

sphuber commented on tests/cli/test_install.py:
```diff
@@ -139,14 +160,15 @@ def test_install_family_url(run_cli_command):
-    assert family.__class__ is SsspFamily
+    family = PseudoPotentialFamily.objects.get(label=label)
+    assert isinstance(family.pseudos['Si'], UpfData)
+    #assert family.__class__ is SsspFamily
```
```suggestion
    assert family.__class__ is PseudoPotentialFamily
```

sphuber commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,43 +24,56 @@ def cmd_install():
 @options_core.DESCRIPTION(help='Description for the family.')
 @options.ARCHIVE_FORMAT()
-@options.FAMILY_TYPE()
+@options.PSEUDOS_TYPE()
```
```suggestion
@options.PSEUDO_TYPE()
```

sphuber commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,43 +24,56 @@ def cmd_install():
-    except AttributeError:
-        suffix = os.path.basename(archive.name)
-
-    # Write the content of the archive to a temporary file, because `create_family_from_archive` does currently not
-    # accept filelike objects because the underlying `shutil.unpack_archive` does not. Likewise, `unpack_archive` will
-    # attempt to deduce the archive format from the filename extension, so it is important we maintain the original
-    # filename. Of course if this fails, users can specify the archive format explicitly wiht the corresponding option.
```
Please reinstate this comment as it is not obvious why we need to preserve the suffix

sphuber commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,43 +24,56 @@ def cmd_install():
-        handle.flush()
+    # `archive_or_folder` is a simple string, containing the name of the folder / file / url location.
 
+    if os.path.isdir(archive_or_folder):
```
We should start using `pathlib` exclusively for filepath manipulations.

sphuber commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,43 +24,56 @@ def cmd_install():
+            )
+    else:
+        # The file of the url must be copied to a local temporary file. Maybe better ways to do it?
+        with urllib.request.urlopen(archive_or_folder) as archive:
```
You are not checking if the opening of the URL actually works and so this will just except if there is anything wrong (I realize the original code also didn't check but it would be good to add now). Should probably use the `requests` library instead which makes it easy to check for problems. See the `download_sssp` for an example.

bosonie replied:
This is already checked by `PathOrUrl` I believe.

sphuber commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,43 +24,56 @@ def cmd_install():
-    The command will attempt to infer the archive format from the filename extension of the ARCHIVE. If this fails, the
-    archive format can be specified explicitly with the archive format option, which will also display which formats
-    are supported.
+    The command will attempt first to recognize the passed `archive_or_folder` as a folder in the local system. If not,
```
```suggestion
    The command will attempt first to recognize the passed ARCHIVE_FOLDER as a folder in the local system. If not,
```
Since these docstrings will be rendered by `click` and are mostly intended for users of the CLI, I prefer to keep a reference to the params by their "metavar" which is formatted by `click`. The reason is that for users in the `--help` it will show as `ARCHIVE_OR_FOLDER` and so this doesn't correspond to `archive_folder`. For developers the inconsistency is less bad because they know that the function argument is turned into all-caps for the help string

sphuber reviewed (changes requested):
Thanks @bosonie . I have given it a first pass. I am wondering if we should reuse the `PathOrUrl` from `aiida-core`. It checks whether the remote URL can be fetched but than discards the result. Maybe it would be best to create our own custom type that does everything in one go as well as validation. But I can look at this myself later.

bosonie commented:
> > I'm also noticing now that
> > ```
> > aiida-pseudo install family https://legacy-archive.materialscloud.org/file/2018.0001/v4/SSSP_1.0_PBE_efficiency.tar.gz label
> > ```
> > 
> > 
> > that should be tested in `tests/cli/test_install.py` does actually fail in real life. Not because of a problem in the CLI command, but for failing of the underline `family_type.create_from_folder`. Is it expected?
> 
> How is it failing exactly? And is it failing on the `master` branch or with your patch? Or both?

```
Info: unpacking archive and parsing pseudos...  [FAILED]
Critical: `<class 'aiida_pseudo.data.pseudo.pseudo.PseudoPotentialData'>` constructor did not define the element and could not parse a valid element symbol from the filename `{filename}` either. It should have the format `ELEMENT.EXTENSION`
```
Also in master

sphuber commented:
> Also in master

I see, well the problem should actually be solved by your PR. Try to run the following on your branch:
```
aiida-pseudo install family https://legacy-archive.materialscloud.org/file/2018.0001/v4/SSSP_1.0_PBE_efficiency.tar.gz label -P pseudo.upf
```
As stated in the docstring, the base family type `PseudoPotentialFamily` will use the basic `PseudoPotentialData` for the pseudopotentials. The `PseudoPotentialData` is a base class for "any" format, but it still needs an element. For now, the way I coded it, it only tries to parse this from the filename, because the file content could be anything and there is no way how to parse it for the element. So we impose a strict naming scheme for the filename `{ELEMENT}.extension`. If this is not the case, we cannot parse the element and so this error is raised. Apparently the `SSSP_1.0_PBE_efficiency.tar.gz` archive contains a file that doesn't follow this convention and so the element cannot be parsed, hence the error. I am a bit hesitant to start making the parsing more clever because there will always be edge cases that we won't catch and this CLI command was supposed to be just a last way out. One can always use the API to directly to create a family however they want.

bosonie commented:
> I see, well the problem should actually be solved by your PR. Try to run the following on your branch:
Indeed!

bosonie commented:
I applied all the changes a part form the one I commented.
I did not created our own custom so far. @sphuber let me know if you wish me to look into it.
Ideally I would love to merge this soon and have a release as well. I will use this features in `aiida-siesta` and I'm close to a release there as well.

sphuber commented:
> Ideally I would love to merge this soon and have a release as well. I will use this features in `aiida-siesta` and I'm close to a release there as well.

Just out of curiosity, since this PR just changes the CLI, why would you need this for `aiida-siesta`? And there are still a few unaddressed comments. I will comment on those specifically so you know which ones.

bosonie commented:
> Just out of curiosity, since this PR just changes the CLI, why would you need this for `aiida-siesta`? And there are still a few unaddressed comments. I will comment on those specifically so you know which ones.

Because in the package we deliver also examples and material for a tutorial. They require the creation of pseudo families and I do not want to make the topic more complicated than necessary.

sphuber commented:
> Because in the package we deliver also examples and material for a tutorial. They require the creation of pseudo families and I do not want to make the topic more complicated than necessary.

But you guys use psml right, and pseudos from Pseudo Dojo? So why can't you use `aiida-pseudo install pseudo-dojo`?

bosonie commented:
> But you guys use psml right, and pseudos from Pseudo Dojo? So why can't you use `aiida-pseudo install pseudo-dojo`?

No, `psml` is a new addition and will be the standard in the future. The current "stable" version of siesta still not support `psml`, only `psf`.

mbercx commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,43 +24,60 @@ def cmd_install():
 @options_core.DESCRIPTION(help='Description for the family.')
 @options.ARCHIVE_FORMAT()
-@options.FAMILY_TYPE()
+@options.PSEUDO_TYPE()
```
I'm not sure if we should remove `FAMILY_TYPE` input here. I was trying to install a family as a `CutoffsFamily`, since the basic `PseudoPotentialFamily` class doesn't have the `RecommendedCutoffMixin` as a parent class, and I wanted to add recommended cutoffs to be used with the `get_builder_from_protocol()` method.

mbercx replied:
Also related to #78.

mbercx reviewed (commented):
Thanks @bosonie! I tried it out on a folder with pseudos I actually needed for a project, so you saved me the trouble of putting them in an archive! ;)

I just had one comment regarding removing the `FAMILY_TYPE` input. Although I do understand that if someone wants to install an `SsspFamily` or `PseudoDojoFamily`, they can use the corresponding command, I actually just now noticed that the base `PseudoPotentialFamily` doesn't have the `RecommendedCutoffMixin` as a parent class. So any pseudos installed with the command currently cannot use this functionality. There is the `CutoffsFamily` class, which @sphuber added mainly for testing purposes, but I'm thinking that perhaps we should make it possible for users to install this class type with the `install family` command. Or, is there any reason _not_ to give the base `PseudoPotentialFamily` class the methods for adding and using recommended cutoffs?

sphuber commented:
@bosonie I have added a commit that implements our own version of `PathOrUrl` that makes the behavior of the command more consistent with respect to error handling. I have updated and simplified the implementation of the `install family` command accordingly and. If you are ok with the change, I can merge this. One more thing I realized while working on this, is that we probably should keep the option to change the family type. Creating an `SsspFamily` doesn't make sense, but I have added the `CutoffsFamily` and this will allow one to then add custom recommended cutoffs which may be useful. I will add this now

sphuber commented:
Haha, guess @mbercx agrees with me on that last part. I am not sure if we should add the `RecommendedCutoffMixin` to the `PseudoPotentialFamily` base class. Especially since we already have the `CutoffsFamily`, which indeed I mainly added for testing purposes, but now with the fully fledged `install family` command, it starts to make sense for real use. Maybe we should rename it slightly though to `CutoffsPseudoPotentialFamily` to still make it plainly obvious that it is a pseudopotential family. I would also still check the type and only allow `pseudo.family` and `pseudo.family.cutoffs`. I don't think we want to allow creating pseudo-dojo or SSSP families through this command. Those respective commands already do that and if there are problems with internet, they even have the option to install directly from archive on disk.

sphuber commented:
@bosonie and @mbercx added the implementation for the family type. Please have a look

mbercx commented:
> Haha, guess @mbercx agrees with me on that last part. 

😁 

> I am not sure if we should add the `RecommendedCutoffMixin` to the `PseudoPotentialFamily` base class. 

Sure, I do appreciate the cleanliness of a mixin class. I'm just wondering if there is a use case for a pseudo family _without_ recommended cutoffs. There probably is, I'm just too in ❤️  with the `get_builder_from_protocol` method.

> Especially since we already have the `CutoffsFamily`, which indeed I mainly added for testing purposes, but now with the fully fledged `install family` command, it starts to make sense for real use. Maybe we should rename it slightly though to `CutoffsPseudoPotentialFamily` to still make it plainly obvious that it is a pseudopotential family. 

Fully agree 👍 

> I would also still check the type and only allow `pseudo.family` and `pseudo.family.cutoffs`. I don't think we want to allow creating pseudo-dojo or SSSP families through this command. Those respective commands already do that and if there are problems with internet, they even have the option to install directly from archive on disk.

Although I do agree, I did have a use case for installing a different version of the SSSP today, since there is a small issue with the header of the lanthanide SSSP files and I wanted to test the fixed versions. To circumvent the restriction on the label, I installed the fixed version as `SSSP/1.0/efficiency` 😬 . Anyways, this is a rare developer use case, and can be easily accomplished by installing a `CutoffsPseudoPotentialFamily` in the future.

mbercx reviewed (approved):
Looks good, thanks @sphuber! Just a small comment that we should perhaps help the user to the right command if they want to install `sssp` or `pseudo_dojo` families.

Also: Maybe we should immediately adapt the docstring of `CutoffsFamily` now that it is a proper class with use cases beyond testing? :)

sphuber commented on aiida_pseudo/cli/params/types.py:
```diff
@@ -78,6 +86,9 @@ def convert(self, value, _, __):
             raise click.BadParameter(f'`{value}` is not an existing group plugin.') from exception
 
+        if self.exclude and value in self.exclude:
+            raise click.BadParameter(f'`{value}` is not an accepted value for this option.')
```
Just would be tricky to do that here. Sure now the type is used only in one command and there I exclude two values that each have their own dedicated command, but that might not always apply. So putting that logic here feels too much like hard-coding. I think we should rather just add this to the README/documentation.

mbercx replied:
Yeah, I see your point. I guess I better get to work on that documentation! 😅

sphuber commented:
OK, we are starting to cram a bit much into a single PR. I am going to factor out the recent additions and then rebase this so it includes mostly the original idea of fixing `install family`.

bosonie commented:
All good from my side! Unlike @mbercx, I was always familiar with families without recommended cutoffs. The good side of a collaborative effort!
Thanks a million!

mbercx commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,46 +22,70 @@ def cmd_install():
+    follow the naming of the `shutil.unpack_archive` standard library method.
+
+    Once the ARCHIVE is downloaded, uncompressed and unarchived into a directory on the local file system, the command
+    will create a `PseudoPotentialFamily` instance where the type of the pseudopotential data nodes that are stored
```
Now that the `FAMILY_TYPE` option has been reinstated, we should probably also revert back to explaining that the pseudo family instance class can be set.

mbercx commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,46 +22,70 @@ def cmd_install():
+    Once the ARCHIVE is downloaded, uncompressed and unarchived into a directory on the local file system, the command
+    will create a `PseudoPotentialFamily` instance where the type of the pseudopotential data nodes that are stored
+    within it is set through the `-P/--pseudo-type` option. If the default `pseudo` (which corresponds to the data
+    plugin `PseudoPotentialData`) the pseudopotential files in the archive *have* to have filenames that strictly follow
```
```suggestion
    plugin `PseudoPotentialData`) is used, the pseudopotential files in the archive *have* to have filenames that strictly follow
```

mbercx commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,46 +22,70 @@ def cmd_install():
+    else:
+        # At this point, we can assume that it is not a valid filepath on disk, but rather a URL and the ``archive``
+        # variable will contain the result objects from the ``requests`` library. The validation of the URL will already
+        # have been done by the ``PathOrUrl`` parameter type, so the URL is reachabel. The content of the URL must be
```
```suggestion
        # have been done by the ``PathOrUrl`` parameter type, so the URL is reachable. The content of the URL must be
```

mbercx commented on aiida_pseudo/cli/install.py:
```diff
@@ -23,46 +22,70 @@ def cmd_install():
 
     family.description = description
-    echo.echo_success(f'installed `{label}` containing {family.count()} pseudo potentials')
+    echo.echo_success(f'installed `{label}` containing {family.count()} pseudopotentials')
```
![image](https://user-images.githubusercontent.com/25607960/116966481-92ad8900-acb0-11eb-90f8-a7b9d2400c29.png)

sphuber replied:
I learn from the best

mbercx reviewed (approved):
Thanks @sphuber! Just gave the PR a final pass, and have left some minor comments. Merge at your discretion! 

> Unlike @mbercx, I was always familiar with families without recommended cutoffs. 

I guess I'm a little spoiled. 😅

sphuber commented:
Thanks a lot @bosonie !
